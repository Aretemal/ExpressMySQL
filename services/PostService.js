import Post from '../models/Post.js';

class PostService {
  async create(post) {
    const createdPost = await Post.create(post);
    return createdPost;
  }

  async getAll(id) {
    const posts = await Post.findAll({
      where: {
        authorId: id,
      },
    });
    return posts;
  }

  async getOne(id) {
    if (!id) {
      throw new Error('Id не указан');
    }
    const post = await Post.findByPk(id);
    return post;
  }

  async update({ author, title, content }, id) {
    if (!id) {
      throw new Error('Id не указан');
    }
    const oldPost = await Post.findByPk(id);
    const updatedPost = await oldPost.update({ author, title, content });
    return updatedPost;
  }

  async delete(id) {
    if (!id) {
      throw new Error('Id не указан');
    }
    const oldPost = await Post.findByPk(id);
    const deletedPost = await oldPost.destroy();
    return deletedPost;
  }
}

export default new PostService();
