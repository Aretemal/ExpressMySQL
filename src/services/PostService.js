import Post from '../models/post.js';
import Comment from '../models/comment.js';
import LikePost from '../models/likePost.js';

class PostService {
  async create({ newMessageText }, id) {
    const createdPost = await Post.create({ authorId: id, content: newMessageText });
    return createdPost;
  }

  async getAll(id) {
    const posts = await Post.findAll({ where: { authorId: id } });
    const postsWithCommentsAndLikes = await Promise.all(posts.map(async (post) => {
      const likesCount = await LikePost.findAll({ where: { postLikeId: post.id } });
      post.likesCount = likesCount.length;
      const isMeLike = await LikePost.findAll({
        where: {
          postLikeId: post.id,
          senderLikeId: id
        }
      });
      post.isMeLike = !!isMeLike.length;
      const commentCount = await Comment.findAll({ where: { postCommentId: post.id } });
      post.commentCount = commentCount.length;
      return post;
    }));
    return postsWithCommentsAndLikes;
  }

  async getOne(id, next) {
    if (!id) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const post = await Post.findByPk(id);
    return post;
  }

  async update({ author, title, content }, id, next) {
    if (!id) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const oldPost = await Post.findByPk(id);
    const updatedPost = await oldPost.update({ author, title, content });
    return updatedPost;
  }

  async delete(id, next) {
    if (!id) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const oldPost = await Post.findByPk(id);
    const deletedPost = await oldPost.destroy();
    return deletedPost;
  }

  async createComment({ id, message }, senderId, next) {
    if (!id && !message) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const comment = await Comment.create({
      postCommentId: id,
      senderCommentId: senderId,
      message,
    });
    return comment;
  }

  async deleteComment(id, next) {
    if (!id) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const comment = await Comment.findByPk(id);
    const deletedComment = await comment.destroy();
    return deletedComment;
  }

  async getAllComments(id, next) {
    if (!id) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const comments = await Comment.findAll({
      where: {
        postCommentId: id,
      },
    });
    return comments;
  }

  async setLike(postId, authId, next) {
    if (!postId) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const checkLike = await LikePost.findOne({
      where: {
        senderLikeId: authId,
        postLikeId: postId,
      },
    });
    if (checkLike) {
      next({ errorsArray: [{ msg: 'Like defined' }] });
      return;
    }
    const like = await LikePost.create({
      senderLikeId: authId,
      postLikeId: postId,
    });
    return like;
  }

  async deleteLike(postId, authId, next) {
    if (!postId) {
      next({ errorsArray: [{ msg: 'Id not specified' }] });
      return;
    }
    const like = await LikePost.findOne({
      where: {
        postLikeId: postId,
        senderLikeId: authId,
      },
    });
    if (!like) {
      next({ errorsArray: [{ msg: 'Like not found' }] });
      return;
    }
    const deletedLike = await like.destroy();
    return deletedLike;
  }
}

export default new PostService();
