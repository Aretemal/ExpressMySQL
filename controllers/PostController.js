import PostService from '../services/PostService.js';

class PostController {
  async create(req, res) {
    const post = await PostService.create(req.body, req.user.id);
    res.json(post);
  }

  async getAll(req, res) {
    const posts = await PostService.getAll(req.user.id);
    return res.json(posts);
  }

  async getOne(req, res) {
    const post = await PostService.getOne(req.params.id);
    return res.json(post);
  }

  async update(req, res) {
    const { author, title, content } = req.body;
    const updatedPost = await PostService.update({ author, title, content }, req.params.id);
    return res.json(updatedPost);
  }

  async delete(req, res) {
    const oldPost = await PostService.delete(req.params.id);
    return res.json(oldPost);
  }
}
export default new PostController();
