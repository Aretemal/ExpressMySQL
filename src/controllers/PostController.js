import { validationResult } from 'express-validator';
import CollectionSerializer from '../serializers/CollectionSerializer.js';
import PostService from '../services/PostService.js';
import PostSerializer from '../serializers/PostSerializer.js';
import AppError from '../utils/AppError.js';

class PostController {
  async create(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      const errorString = errors.array().map((item) => item.msg).join(', ');
      throw new AppError(errorString, 400);
    }
    const post = await PostService.create(req.body, req.user.id);
    req.serializer = new PostSerializer(post);
    next();
  }

  async getAll(req, res, next) {
    const posts = await PostService.getAll(req.user.id);
    req.serializer = new CollectionSerializer(posts, { serializerType: PostSerializer });
    next();
  }

  async getOne(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      const errorString = errors.array().map((item) => item.msg).join(', ');
      throw new AppError(errorString, 400);
    }
    const post = await PostService.getOne(req.params.id);
    req.serializer = new PostSerializer(post);
    next();
  }

  async update(req, res, next) {
    const { author, title, content } = req.body;
    const updatedPost = await PostService.update({ author, title, content }, req.params.id);
    req.serializer = new PostSerializer(updatedPost);
    next();
  }

  async delete(req, res, next) {
    const oldPost = await PostService.delete(req.params.id);
    req.serializer = new PostSerializer(oldPost);
    next();
  }
}
export default new PostController();
