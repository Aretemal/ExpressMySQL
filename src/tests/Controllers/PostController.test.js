/* eslint-disable no-undef */
import { jest } from '@jest/globals';
import PostController from '../../controllers/PostController';
import PostService from '../../services/PostService';

jest.fn('PostService');
const req = {
  body: {
    newMessageText: 'Hello',
  },
  user: {
    id: 1,
  },
  params: {
    id: 1,
  },
};
const res = {
  dataJS: null,
  json: (data) => {
    res.dataJS = data;
  },
};

describe('Post Controller :', () => {
  describe('Create', () => {
    test('should creates a new post', async () => {
      function next() {}
      PostService.create = ({ newMessageText }, id) => ({
        authorId: id,
        content: newMessageText,
      });

      await PostController.create(req, res, next);

      expect(req.serializer).toMatchSnapshot();
    });
  });

  describe('getAll', () => {
    test('should returns all posts', async () => {
      function next() {}
      PostService.getAll = () => ([
        { id: 1 }, { id: 2 },
      ]);

      await PostController.getAll(req, res, next);

      expect(req.serializer).toMatchSnapshot();
    });
  });

  describe('getOne', () => {
    test('should returns one posts', async () => {
      function next() {}
      PostService.getOne = (id) => ({
        id, content: 'Hello',
      });

      await PostController.getOne(req, res, next);

      expect(req.serializer).toMatchSnapshot();
    });
  });
});
