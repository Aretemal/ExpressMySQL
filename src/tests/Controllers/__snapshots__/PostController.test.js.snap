// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Controller : Create should creates a new post 1`] = `
PostSerializer {
  "resource": {
    "authorId": 1,
    "content": "Hello",
  },
}
`;

exports[`Post Controller : getAll should returns all posts 1`] = `
CollectionSerializer {
  "resource": [
    {
      "id": 1,
    },
    {
      "id": 2,
    },
  ],
  "serializerType": [Function],
}
`;

exports[`Post Controller : getOne should returns one posts 1`] = `
PostSerializer {
  "resource": {
    "content": "Hello",
    "id": 1,
  },
}
`;
